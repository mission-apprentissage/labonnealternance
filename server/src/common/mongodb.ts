import mongoose from "mongoose"
import config from "../config.js"

export let mongooseInstance = mongoose
export const connectToMongo = (mongoUri = config.mongodb.uri, mongooseInst = null) => {
  return new Promise((resolve, reject) => {
    console.log(`MongoDB: Connection to ${mongoUri}`)

    const mI = mongooseInst || mongooseInstance
    // Set up default mongoose connection
    mI.connect(mongoUri, {
      useNewUrlParser: true,
      useUnifiedTopology: true,
      useFindAndModify: false,
      keepAlive: true,
    })

    // Get Mongoose to use the global promise library
    mI.Promise = global.Promise // Get the default connection
    const db = mI.connection

    // Bind connection to error event (to get notification of connection errors)
    db.on("error", (e) => {
      console.error("MongoDB: connection error:")
      reject(e)
    })

    db.once("open", () => {
      console.log("MongoDB: Connected")
      resolve({ db })
    })
  })
}

export const closeMongoConnection = (mongooseInst = mongoose) => mongooseInst.disconnect()
