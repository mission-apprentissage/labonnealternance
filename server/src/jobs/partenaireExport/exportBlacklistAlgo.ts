import { LBA_ITEM_TYPE } from "shared/constants/lbaitem"

import { s3WriteString } from "@/common/utils/awsUtils"
import { getDbCollection } from "@/common/utils/mongodbUtils"
import { notifyToSlack } from "@/common/utils/slackUtils"

const exportBlacklistAlgo = async (): Promise<void> => {
  const reportedCompanies = await getDbCollection("reported_companies")
    .find({ reason: "Entreprise fermée", type: LBA_ITEM_TYPE.RECRUTEURS_LBA }, { projection: { _id: 0, siret: 1 } })
    .toArray()
  const unsubscribedCompanies = await getDbCollection("unsubscribedrecruteurslba")
    .find({}, { projection: { _id: 0, siret: 1 } })
    .toArray()
  const uniqueSiret = [...new Set([...reportedCompanies, ...unsubscribedCompanies].map((company) => company.siret))]
  const jsonContent = JSON.stringify(uniqueSiret, null, 2)
  const fileKey = `blacklist-algo.json`

  await s3WriteString("storage", fileKey, {
    Body: jsonContent,
    ContentType: "application/json",
  })

  await notifyToSlack({ subject: "Fichier de blacklist pour l'algorithme", message: `${uniqueSiret.length} siret exportés` })
}

const exportApplications = async (): Promise<void> => {
  const applications = await getDbCollection("applications")
    .find({ job_origin: LBA_ITEM_TYPE.RECRUTEURS_LBA }, { projection: { _id: 1, company_siret: 1, created_at: 1 } })
    .toArray()
  const jsonContent = JSON.stringify(applications, null, 2)
  const fileKey = `applications-algo.json`
  await s3WriteString("storage", fileKey, {
    Body: jsonContent,
    ContentType: "application/json",
  })

  await notifyToSlack({ subject: "Fichier des candidatures pour l'algorithme", message: `${applications.length} candidatures exportées` })
}

export const exportFileForAlgo = async (): Promise<void> => {
  await exportBlacklistAlgo()
  await exportApplications()
}
