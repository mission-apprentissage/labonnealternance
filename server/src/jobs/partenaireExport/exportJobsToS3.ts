import { writeFileSync } from "fs"

import { LBA_ITEM_TYPE } from "shared/constants/lbaitem"

import { uploadFileToS3 } from "../../common/utils/awsUtils"
import { getDbCollection } from "../../common/utils/mongodbUtils"

interface IGeneratorParams {
  collection: "jobs" | "recruteurslba"
  query: object
  projection: object
  fileName: LBA_ITEM_TYPE.OFFRES_EMPLOI_LBA | LBA_ITEM_TYPE.RECRUTEURS_LBA
}

async function generateJsonExport({ collection, query, projection, fileName }: IGeneratorParams): Promise<string> {
  const filePath = new URL(`./${fileName}.json`, import.meta.url)
  const data = await getDbCollection(collection).find(query).project(projection).toArray()
  writeFileSync(filePath, JSON.stringify(data, null, 4))
  return filePath.pathname
}

async function exportLbaJobsToS3() {
  const offres_emploi_lba: IGeneratorParams = {
    collection: "jobs",
    query: { job_status: "Active" },
    projection: {
      rome_detail: 0,
      _id: 0,
      delegations: 0,
      address_detail: 0,
      jobId: 0,
      recruiterId: 0,
      email: 0,
      phone: 0,
      first_name: 0,
      last_name: 0,
      relance_mail_sent: 0,
      is_disabled_elligible: 0,
      establishment_id: 0,
      origin: 0,
      recruiterStatus: 0,
      is_multi_published: 0,
      job_delegation_count: 0,
      stats_detail_view: 0,
      stats_search_view: 0,
      job_status: 0,
    },
    fileName: LBA_ITEM_TYPE.OFFRES_EMPLOI_LBA,
  }
  const recruteurs_lba: IGeneratorParams = {
    collection: "recruteurslba",
    query: {},
    projection: { _id: 0, email: 0, phone: 0, geopoint: 0, recruitment_potential: 0, opco_short_name: 0 },
    fileName: LBA_ITEM_TYPE.RECRUTEURS_LBA,
  }
  await Promise.all([
    generateJsonExport(offres_emploi_lba).then((path) => uploadFileToS3({ key: path.split("/").pop() as string, filePath: path, noCache: true })),
    generateJsonExport(recruteurs_lba).then((path) => uploadFileToS3({ key: path.split("/").pop() as string, filePath: path, noCache: true })),
  ])
}

export { exportLbaJobsToS3 }
