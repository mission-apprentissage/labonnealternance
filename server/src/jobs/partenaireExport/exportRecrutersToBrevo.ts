import { LBA_ITEM_TYPE } from "shared/constants/lbaitem"

import { getDbCollection } from "@/common/utils/mongodbUtils"
import { uploadContactListToBrevo } from "@/services/brevo.service"

export const exportRecruteursToBrevo = async () => {
  const applications = await getDbCollection("applications")
    .aggregate([
      {
        $match: {
          created_at: { $gte: new Date("2025-09-15") }, // demandé par le métier @aurélie
          job_origin: LBA_ITEM_TYPE.RECRUTEURS_LBA,
          company_recruitment_intention: { $ne: null },
          company_feedback_date: { $ne: null },
        },
      },
      {
        $lookup: {
          from: "applicants",
          localField: "applicant_id",
          foreignField: "_id",
          as: "applicant",
        },
      },
      { $unwind: "$applicant" },
      {
        $project: {
          _id: 0,
          email: "$company_email",
          firstname: "$applicant.firstname",
          lastname: "$applicant.lastname",
          company_feedback: "$company_recruitment_intention",
          feedback_date: "$company_feedback_date",
        },
      },
    ])
    .toArray()

  await uploadContactListToBrevo(
    "MARKETING",
    applications,
    [
      {
        key: "email",
        header: "EMAIL",
      },
      {
        key: "firstname",
        header: "PRENOM",
      },
      {
        key: "lastname",
        header: "NOM",
      },
      {
        key: "company_feedback",
        header: "REPONSE_EMPLOYEUR",
      },
      {
        key: "feedback_date",
        header: "DATE_REPONSE_EMPLOYEUR",
      },
    ],
    "628"
  )
}
