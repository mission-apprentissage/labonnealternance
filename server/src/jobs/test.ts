import { ObjectId } from "bson"
import { IApplicant, IApplication, ZApplicant } from "shared"

import { getDbCollection } from "@/common/utils/mongodbUtils"

import { asyncForEach } from "../common/utils/asyncUtils"

type ApplicationAggregate = Pick<IApplication, "applicant_email" | "applicant_first_name" | "applicant_last_name" | "applicant_phone" | "_id">

export const migrationTest = async () => {
  // console.log(extensions.telephone.safeParse("0690760552"))
  // return
  const applications = (await getDbCollection("applications")
    .find({}, { projection: { applicant_first_name: 1, applicant_last_name: 1, applicant_email: 1, applicant_phone: 1, _id: 1 } })
    .toArray()) as ApplicationAggregate[]

  const stat = { error: 0, success: 0, total: applications.length }

  await asyncForEach(applications, async (application) => {
    const { applicant_first_name, applicant_email, applicant_last_name, applicant_phone, _id } = application
    const now = new Date()
    const applicant: IApplicant = {
      _id: new ObjectId(),
      firstname: applicant_first_name,
      lastname: applicant_last_name,
      phone: applicant_phone,
      email: applicant_email,
      last_connection: null,
      createdAt: now,
      updatedAt: now,
    }
    const validation = ZApplicant.safeParse(applicant)
    if (!validation.success) {
      stat.error++
      console.log({ applicant, err: validation.error.errors })
    } else {
      stat.success++
    }
  })
  console.log(stat)
}
