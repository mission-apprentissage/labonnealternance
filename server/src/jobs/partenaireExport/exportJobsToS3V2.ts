import { createReadStream, createWriteStream } from "fs"
import Stream, { Transform } from "stream"
import { pipeline } from "stream/promises"

import { JOB_STATUS_ENGLISH } from "shared"
import { IJobsPartnersOfferPrivate } from "shared/models/jobsPartners.model"

import { concatStreams, waitForStreamEnd } from "@/common/utils/streamUtils"
import { jobsPartnersToApiV3Read } from "@/services/jobs/jobOpportunity/jobOpportunity.service"

import { logger } from "../../common/logger"
import { s3WriteStream } from "../../common/utils/awsUtils"
import { getDbCollection } from "../../common/utils/mongodbUtils"

const getFileWriteStream = (fileName: string) => {
  logger.info(`Generating file ${fileName}`)
  const url = new URL(`./${fileName}`, import.meta.url)
  const writeStream = createWriteStream(url)
  return { writeStream, filePath: url.pathname }
}

const uploadFileToS3 = async (filepath: string, readStream: Stream.Readable) => {
  const key = filepath.split("/").pop() as string
  logger.info(`Uploading file ${filepath} to S3`)
  await s3WriteStream("storage", key, { Body: readStream, CacheControl: "no-cache, no-store, must-revalidate" })
  logger.info(`file ${filepath} uploaded`)
}

export const EXPORT_JOBS_TO_S3_V2_FILENAME = "exportJobsToS3V2.json"

export async function exportJobsToS3V2(handleFileReadStream = uploadFileToS3) {
  const { filePath, writeStream: fileStream } = getFileWriteStream(EXPORT_JOBS_TO_S3_V2_FILENAME)
  const jsonArrayTransform = getJsonArrayTransform()
  jsonArrayTransform.setMaxListeners(20)

  const jobPartnersTransform = getCusorTransform<IJobsPartnersOfferPrivate>((jobPartner) => {
    return JSON.stringify(jobsPartnersToApiV3Read(jobPartner), null, 2)
  })

  concatStreams([jobPartnersTransform], jsonArrayTransform)
  jsonArrayTransform.pipe(fileStream, { end: true })

  logger.info("starting job partners cursor")
  const jobPartnersCursor = await getDbCollection("jobs_partners").find({ offer_status: JOB_STATUS_ENGLISH.ACTIVE, offer_multicast: true })
  await pipeline(jobPartnersCursor, jobPartnersTransform)

  logger.info("waiting for jsonArrayTransform")
  await waitForStreamEnd(jsonArrayTransform)
  console.log("waiting for fileStream")
  await waitForStreamEnd(fileStream)

  logger.info("creating read stream")
  const fileReadStream = createReadStream(filePath)
  await handleFileReadStream(filePath, fileReadStream)
}

function getJsonArrayTransform() {
  let isFirst = true
  const transform = new Transform({
    objectMode: true,
    transform(chunk, encoding, callback) {
      if (!chunk) {
        callback()
        return
      }
      if (isFirst) {
        this.push("[")
        isFirst = false
      } else {
        this.push(",")
      }
      this.push(chunk)
      callback()
    },
    flush(callback) {
      if (isFirst) {
        this.push("[")
        isFirst = false
      }
      this.push("]")
      callback()
    },
  })
  return transform
}

const getCusorTransform = <T>(mapper: (obj: T) => string | null) => {
  return new Transform({
    readableObjectMode: true,
    writableObjectMode: true,
    transform(chunk: T, encoding, callback) {
      const result = mapper(chunk)
      if (result !== null) {
        this.push(result)
      }
      callback()
    },
    flush(callback) {
      callback()
    },
  })
}
