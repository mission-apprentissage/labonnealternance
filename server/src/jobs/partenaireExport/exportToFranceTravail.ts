import { createWriteStream } from "fs"
import path from "path"
import { Readable } from "stream"

import { oleoduc, transformData, transformIntoCSV } from "oleoduc"
import { RECRUITER_STATUS } from "shared/constants/recruteur"
import { JOB_STATUS } from "shared/models"

import { db } from "@/common/mongodb"

import { sendCsvToFranceTravail } from "../../common/apis/FranceTravail"
import { logger } from "../../common/logger"
import { Cfa } from "../../common/model/index"
import { getDepartmentByZipCode } from "../../common/territoires"
import { asyncForEach } from "../../common/utils/asyncUtils"
import { notifyToSlack } from "../../common/utils/slackUtils"
import dayjs from "../../services/dayjs.service"

const regex = /^(.*) (\d{4,5}) (.*)$/
const formatDate = (date) => dayjs(date).format("DD/MM/YYYY")
const splitter = (str) => str.split(regex).filter(String)

/**
 * @description format data into Pole Emploi specific fields
 * @param {object} offre
 * @returns {Promise<object>}
 */
const formatToPe = async (offre) => {
  //Récupération de l'appellation dans le rome_detail pour identifier le code OGR
  const appellation = offre.rome_detail.appellations.find((v) => v.libelle === offre.rome_appellation_label)
  const adresse = offre.address_detail
  const [latitude, longitude] = offre.geo_coordinates.split(",")

  const [rue, code_postal, ville] = offre.is_delegated && offre.cfa?.address_detail?.label ? splitter(offre.cfa.address_detail.label) : []

  const ntcCle = offre.type === "Apprentissage" ? "E2" : "FS"

  if (!appellation || !adresse || !latitude || !longitude) {
    return
  }

  return {
    Par_ref_offre: `${ntcCle}-${offre.jobId}`,
    Par_cle: "LABONNEALTERNANCE",
    Par_nom: "LABONNEALTERNANCE",
    Par_URL_offre: `https://labonnealternance.apprentissage.beta.gouv.fr/recherche-apprentissage?&type=matcha&itemId=${offre.jobId}`,
    Code_rome: offre.rome_code[0],
    Code_OGR: appellation.code_ogr,
    Libelle_metier_OGR: appellation.libelle,
    Description: offre.rome_detail.definition,
    Off_experience_duree_min: null,
    Off_experience_duree_max: null,
    Exp_cle: "D",
    Exp_libelle: null,
    Dur_cle_experience: null,
    Dur_libelle_experience: null,
    Off_experience_commentaire: null,
    Qua_cle: null,
    Qua_libelle: null,
    SCN_cle: offre.naf_code,
    SCN_libelle: offre.naf_label,
    Tfm_cle_1: null,
    Tfm_libelle_1: null,
    Dfm_cle_1: null,
    Dfm_libelle_1: null,
    Dfm_exi_cle_1: null,
    Dfm_exi_libelle_1: null,
    Tfm_cle_2: null,
    Tfm_libelle_2: null,
    Dfm_cle_2: null,
    Dfm_libelle_2: null,
    Dfm_exi_cle_2: null,
    Dfm_exi_libelle_2: null,
    Lan_cle_1: null,
    Lan_libelle_1: null,
    NVL_cle_1: null,
    NVL_libelle_1: null,
    Lan_exi_cle_1: null,
    Lan_exi_libelle_1: null,
    Lan_cle_2: null,
    Lan_libelle_2: null,
    NVL_cle_2: null,
    NVL_libelle_2: null,
    Lan_exi_cle_2: null,
    Lan_exi_libelle_2: null,
    TSA_cle: null,
    TSA_libelle: null,
    Off_salaire_min: null,
    Off_salaire_max: null,
    UMO_cle: null,
    UMO_libelle: null,
    Off_salaire_nb_mois: null,
    Off_salaire_cpt_commentaire: null,
    Off_travail_hebdo_nb_hh: null,
    Off_travail_hebdo_nb_mi: null,
    THO_cle: null,
    THO_libelle: null,
    Off_THO_commentaire: null,
    NTC_cle: ntcCle,
    NTC_libelle: null,
    TCO_cle: "CDD",
    TCO_libelle: null,
    Off_contrat_duree_MO: offre.job_duration,
    Off_contrat_duree_JO: null,
    Off_adr_id: null,
    Off_adr_norme: null,
    Off_adr_compl_1: null,
    Off_adr_compl_2: null,
    Off_adr_no_voie: adresse.numero_voie,
    Off_adr_nom_voie: `${adresse.type_voie ?? ""} ${adresse.nom_voie ?? ""}`,
    CPO_cle: null,
    COM_cle: adresse.code_insee_localite,
    COM_libelle: null,
    DEP_cle: adresse.code_postal ? adresse.code_postal.slice(0, 2) : "",
    DEP_libelle: null,
    REG_cle: adresse.code_postal ? getDepartmentByZipCode(adresse.code_postal)?.region.code ?? "" : "",
    REG_libelle: null,
    Pay_cle: null,
    Pay_libelle: adresse.l7,
    CON_cle: null,
    CON_libelle: null,
    Coordonnee_geo_loc_1: latitude,
    Coordonnee_geo_loc_2: longitude,
    PCO_cle_1: null,
    PCO_libelle_1: null,
    PCO_exi_cle_1: null,
    PCO_exi_libelle_1: null,
    PCO_cle_2: null,
    PCO_libelle_2: null,
    PCO_exi_cle_2: null,
    PCO_exi_libelle_2: null,
    Off_date_creation: formatDate(offre.job_creation_date),
    Off_date_modification: formatDate(offre.job_update_date),
    OST_poste_restant_nb: offre.job_count,
    Off_client_final_siret: offre.establishment_siret,
    Off_client_final_nom: adresse.l7,
    Off_etab_enseigne: offre.cfa ? offre.cfa?.establishment_raison_sociale : offre.establishment_raison_sociale ?? null,
    Col_cle: null,
    Col_nom: null,
    Col_URL_offre: null,
    Version: null,
    Type_mouvement: "C",
    Date_debut_contrat: formatDate(offre.job_start_date),
    Motif_suppression: null,
    Description_entreprise: offre.job_description ?? null,
    Off_etab_siret: offre.cfa_delegated_siret,
    Id_recruteur: null,
    Civ_correspondant: null,
    Nom_correspondant: null,
    Prenom_correspondant: null,
    Tel_correspondant: null,
    Mail_correspondant: null,
    Libelle_etab: null,
    Num_voie_etab: null,
    Type_voie_etab: rue ?? null,
    Lib_voie_etab: rue ?? null,
    Cplt_adresse_1: ville,
    Cplt_adresse_2: null,
    Code_postal_etab: code_postal,
    Code_commune_etab: null,
    Service: null,
    Mode_diffusion: "O",
    Rappel: null,
    Mode_presentation: null,
    Emploi_metier_isco: null,
  }
}

/**
 * @description Generate a CSV with eligible offers for Pole Emploi integration
 */
export const exportToFranceTravail = async (): Promise<void> => {
  try {
    const csvPath = new URL("./exportFT.csv", import.meta.url)
    const buffer: any[] = []
    const threshold = dayjs().subtract(30, "days").toDate()

    // Retrieve only active offers
    const offres: any[] = await db
      .collection("jobs")
      .aggregate([
        {
          $match: {
            job_status: JOB_STATUS.ACTIVE,
            recruiterStatus: RECRUITER_STATUS.ACTIF,
            geo_coordinates: { $nin: ["NOT FOUND", null] },
            job_update_date: { $gte: threshold },
            address_detail: { $ne: null },
            is_multi_published: true,
          },
        },
        {
          $lookup: {
            from: "referentielromes",
            localField: "rome_code.0",
            foreignField: "rome.code_rome",
            as: "referentielrome",
          },
        },
        {
          $unwind: { path: "$referentielrome" },
        },
        {
          $set: { rome_detail: "$referentielrome" },
        },
        {
          $project: { referentielrome: 0, "rome_detail._id": 0 },
        },
      ])
      .toArray()

    logger.info(`get info from ${offres.length} offers...`)
    await asyncForEach(offres, async (offre) => {
      const cfa = offre.is_delegated ? await Cfa.findOne({ siret: offre.cfa_delegated_siret }) : null
      offre.job_type.map((type) => {
        const cfaFields = cfa ? { address_detail: cfa.address_detail, establishment_raison_sociale: cfa.raison_sociale } : null
        buffer.push({ ...offre, type, cfa: cfaFields })
      })
    })

    logger.info("Start stream to CSV...")
    await oleoduc(
      Readable.from(buffer),
      transformData((value) => formatToPe(value)),
      transformIntoCSV({ separator: "|" }),
      createWriteStream(csvPath)
    )

    await sendCsvToFranceTravail(path.resolve(csvPath.pathname))

    await notifyToSlack({
      subject: "EXPORT FRANCE TRAVAIL OK",
      message: `${buffer.length} offres transmises à France Travail`,
    })
  } catch (err) {
    await notifyToSlack({
      subject: "EXPORT FRANCE TRAVAIL KO",
      message: `Echec de l'export des offres France Travail. ${err}`,
    })
    throw err
  }
}
