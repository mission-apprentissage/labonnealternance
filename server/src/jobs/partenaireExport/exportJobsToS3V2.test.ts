import fs from "node:fs/promises"
import Stream from "stream"

import { ObjectId } from "mongodb"
import { JOB_STATUS_ENGLISH } from "shared"
import { beforeEach, describe, expect, it, vi } from "vitest"

import { getDbCollection } from "@/common/utils/mongodbUtils"
import { streamToString } from "@/common/utils/streamUtils"
import { EXPORT_JOBS_TO_S3_V2_FILENAME, exportJobsToS3V2 } from "@/jobs/partenaireExport/exportJobsToS3V2"
import { createJobPartner } from "@tests/utils/jobsPartners.test.utils"
import { useMongo } from "@tests/utils/mongo.test.utils"

import { generateJobsPartnersFull } from "../../../../shared/src/fixtures/jobPartners.fixture"

useMongo()

const exportJobsToS3V2ForTests = async () => {
  let content = ""
  const handleStream = async (filepath: string, stream: Stream.Readable) => {
    content = await streamToString(stream)
  }
  await exportJobsToS3V2(handleStream)
  return JSON.parse(content)
}

describe("exportJobsToS3V2", () => {
  beforeEach(async () => {
    vi.useFakeTimers()
    vi.setSystemTime(new Date("2024-08-21"))

    return async () => {
      vi.useRealTimers()
      await getDbCollection("jobs_partners").deleteMany({})
      const filepath = new URL(`./${EXPORT_JOBS_TO_S3_V2_FILENAME}`, import.meta.url).pathname
      try {
        await fs.unlink(filepath)
      } catch (err) {
        console.warn(`could not delete ${filepath}`)
      }
    }
  })

  it("should publish jobs", async () => {
    // given
    await createJobPartner(generateJobsPartnersFull({ _id: new ObjectId("66c52e00c8f12f66f971106f"), partner_job_id: "id_1" }))
    await createJobPartner(generateJobsPartnersFull({ _id: new ObjectId("66c52e00393b331450d8bedc"), partner_job_id: "id_2" }))
    // when
    const jobs = await exportJobsToS3V2ForTests()
    // then
    expect.soft(jobs.length).toBe(2)
    expect.soft(jobs).toMatchSnapshot()
  })
  it("should publish a partenariat job", async () => {
    // given
    await createJobPartner(
      generateJobsPartnersFull({
        _id: new ObjectId("66c52e00c8f12f66f971106f"),
        partner_job_id: "id_1",
        is_delegated: true,
        cfa_siret: "12345678901234",
        cfa_address_label: "rue du cfa 75002 Paris",
        cfa_apply_email: "cfa@cfa.fr",
        cfa_apply_phone: "0123456789",
        cfa_legal_name: "nom CFA",
      })
    )
    // when
    const jobs = await exportJobsToS3V2ForTests()
    // then
    expect.soft(jobs.length).toBe(1)
    expect.soft(jobs).toMatchSnapshot()
  })
  it("should not publish a job partner that is not active", async () => {
    // given
    await createJobPartner(generateJobsPartnersFull({ offer_status: JOB_STATUS_ENGLISH.ANNULEE }))
    await createJobPartner(generateJobsPartnersFull({ offer_status: JOB_STATUS_ENGLISH.POURVUE }))
    // when
    const jobs = await exportJobsToS3V2ForTests()
    // then
    expect.soft(jobs.length).toBe(0)
  })
})
