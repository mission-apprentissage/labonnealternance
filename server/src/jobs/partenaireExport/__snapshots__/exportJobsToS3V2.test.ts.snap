// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`exportJobsToS3V2 > should publish jobs 1`] = `
[
  {
    "apply": {
      "phone": "0123456789",
      "recipient_id": "partners_66c52e00c8f12f66f971106f",
      "url": "https://apply.com",
    },
    "contract": {
      "duration": 6,
      "remote": "onsite",
      "start": "2021-01-28T15:00:00.000Z",
      "type": [
        "Apprentissage",
        "Professionnalisation",
      ],
    },
    "identifier": {
      "id": "66c52e00c8f12f66f971106f",
      "partner_job_id": "id_1",
      "partner_label": "Hellowork",
    },
    "is_delegated": false,
    "offer": {
      "access_conditions": [
        "offer_access_conditions",
      ],
      "description": "Attention il te faut une super motivation pour ce job",
      "desired_skills": [
        "offer_desired_skills",
      ],
      "opening_count": 1,
      "publication": {
        "creation": "2021-01-28T15:00:00.000Z",
        "expiration": "2050-01-01T00:00:00.000Z",
      },
      "rome_codes": [
        "D1102",
      ],
      "status": "Active",
      "target_diploma": {
        "european": "3",
        "label": "Master",
      },
      "title": "Une super offre d'alternance",
      "to_be_acquired_skills": [
        "offer_to_be_acquired_skills",
      ],
    },
    "workplace": {
      "brand": "workplace_brand",
      "description": "workplace_description",
      "domain": {
        "idcc": 1003,
        "naf": {
          "code": "6201Z",
          "label": "Programmation informatique",
        },
        "opco": "ATLAS",
      },
      "legal_name": "workplace_legal_name",
      "location": {
        "address": "126 RUE DE L'UNIVERSITE 75007 PARIS",
        "geopoint": {
          "coordinates": [
            0,
            0,
          ],
          "type": "Point",
        },
      },
      "name": "workplace_name",
      "siret": "13002526500013",
      "size": "0-50",
      "website": "http://workplace_website.com",
    },
  },
  {
    "apply": {
      "phone": "0123456789",
      "recipient_id": "partners_66c52e00393b331450d8bedc",
      "url": "https://apply.com",
    },
    "contract": {
      "duration": 6,
      "remote": "onsite",
      "start": "2021-01-28T15:00:00.000Z",
      "type": [
        "Apprentissage",
        "Professionnalisation",
      ],
    },
    "identifier": {
      "id": "66c52e00393b331450d8bedc",
      "partner_job_id": "id_2",
      "partner_label": "Hellowork",
    },
    "is_delegated": false,
    "offer": {
      "access_conditions": [
        "offer_access_conditions",
      ],
      "description": "Attention il te faut une super motivation pour ce job",
      "desired_skills": [
        "offer_desired_skills",
      ],
      "opening_count": 1,
      "publication": {
        "creation": "2021-01-28T15:00:00.000Z",
        "expiration": "2050-01-01T00:00:00.000Z",
      },
      "rome_codes": [
        "D1102",
      ],
      "status": "Active",
      "target_diploma": {
        "european": "3",
        "label": "Master",
      },
      "title": "Une super offre d'alternance",
      "to_be_acquired_skills": [
        "offer_to_be_acquired_skills",
      ],
    },
    "workplace": {
      "brand": "workplace_brand",
      "description": "workplace_description",
      "domain": {
        "idcc": 1003,
        "naf": {
          "code": "6201Z",
          "label": "Programmation informatique",
        },
        "opco": "ATLAS",
      },
      "legal_name": "workplace_legal_name",
      "location": {
        "address": "126 RUE DE L'UNIVERSITE 75007 PARIS",
        "geopoint": {
          "coordinates": [
            0,
            0,
          ],
          "type": "Point",
        },
      },
      "name": "workplace_name",
      "siret": "13002526500013",
      "size": "0-50",
      "website": "http://workplace_website.com",
    },
  },
  {
    "apply": {
      "phone": null,
      "recipient_id": "recruiters_66c52e0088651f3c2cd0d785",
      "url": "http://localhost:3000/emploi/offres_emploi_lba/66c52e0088651f3c2cd0d785/Boulanger",
    },
    "contract": {
      "duration": null,
      "remote": null,
      "start": "2021-01-28T15:00:00.000Z",
      "type": [
        "Apprentissage",
      ],
    },
    "identifier": {
      "id": "66c52e0088651f3c2cd0d785",
      "partner_job_id": "66c52e0088651f3c2cd0d785",
      "partner_label": "offres_emploi_lba",
    },
    "is_delegated": false,
    "offer": {
      "access_conditions": [
        "Ce métier est accessible avec un diplôme de fin d'études secondaires (brevet des collèges) à Bac (professionnel, Brevet Professionnel, ...) dans le secteur tertiaire. Il est également accessible avec une expérience professionnelle sans diplôme particulier. La maîtrise de l'outil bureautique (traitement de texte, tableur, ...) peut être requise.",
      ],
      "description": "Exécute des travaux administratifs courants (vérification de documents, frappe et mise en forme de courriers pré-établis, suivi de dossier administratifs, ...) selon l'organisation de la structure ou du service. Peut être en charge d'activités de reprographie et d'archivage. Peut réaliser l'accueil de la structure.",
      "desired_skills": [
        "Faire preuve de rigueur et de précision",
        "Organiser son travail selon les priorités et les objectifs",
        "Être à l'écoute, faire preuve d'empathie",
      ],
      "opening_count": 1,
      "publication": {
        "creation": null,
        "expiration": null,
      },
      "rome_codes": [
        "M1602",
      ],
      "status": "Active",
      "target_diploma": null,
      "title": "Boulanger",
      "to_be_acquired_skills": [
        "Production, Fabrication: Procéder à l'enregistrement, au tri, à l'affranchissement du courrier",
        "Production, Fabrication: Réaliser des travaux de reprographie",
        "Gestion des stocks: Contrôler l'état des stocks",
        "Gestion des stocks: Définir des besoins en approvisionnement",
        "Logistique: Organiser le traitement des commandes",
        "Relation client: Accueillir, orienter, informer une personne",
        "Organisation: Contrôler la conformité des données ou des documents",
        "Organisation: Corriger et mettre en forme un document",
        "Organisation: Numériser des documents, médias ou supports techniques",
        "Organisation: Utiliser les outils bureautiques",
        "Organisation: Établir, mettre à jour un dossier, une base de données",
      ],
    },
    "workplace": {
      "brand": null,
      "description": null,
      "domain": {
        "idcc": 1000,
        "naf": null,
        "opco": "ATLAS",
      },
      "legal_name": null,
      "location": {
        "address": "3 rue du loup 75001 Paris",
        "geopoint": {
          "coordinates": [
            0,
            0,
          ],
          "type": "Point",
        },
      },
      "name": null,
      "siret": "11000001500013",
      "size": null,
      "website": null,
    },
  },
  {
    "apply": {
      "phone": null,
      "recipient_id": "recruiters_66c52e00f03faa7ce1aaccbc",
      "url": "http://localhost:3000/emploi/offres_emploi_lba/66c52e00f03faa7ce1aaccbc/Patissier",
    },
    "contract": {
      "duration": null,
      "remote": null,
      "start": "2021-01-28T15:00:00.000Z",
      "type": [
        "Apprentissage",
      ],
    },
    "identifier": {
      "id": "66c52e00f03faa7ce1aaccbc",
      "partner_job_id": "66c52e00f03faa7ce1aaccbc",
      "partner_label": "offres_emploi_lba",
    },
    "is_delegated": false,
    "offer": {
      "access_conditions": [
        "Ce métier est accessible avec un diplôme de fin d'études secondaires (brevet des collèges) à Bac (professionnel, Brevet Professionnel, ...) dans le secteur tertiaire. Il est également accessible avec une expérience professionnelle sans diplôme particulier. La maîtrise de l'outil bureautique (traitement de texte, tableur, ...) peut être requise.",
      ],
      "description": "Exécute des travaux administratifs courants (vérification de documents, frappe et mise en forme de courriers pré-établis, suivi de dossier administratifs, ...) selon l'organisation de la structure ou du service. Peut être en charge d'activités de reprographie et d'archivage. Peut réaliser l'accueil de la structure.",
      "desired_skills": [
        "Faire preuve de rigueur et de précision",
        "Organiser son travail selon les priorités et les objectifs",
        "Être à l'écoute, faire preuve d'empathie",
      ],
      "opening_count": 1,
      "publication": {
        "creation": null,
        "expiration": null,
      },
      "rome_codes": [
        "M1602",
      ],
      "status": "Active",
      "target_diploma": null,
      "title": "Patissier",
      "to_be_acquired_skills": [
        "Production, Fabrication: Procéder à l'enregistrement, au tri, à l'affranchissement du courrier",
        "Production, Fabrication: Réaliser des travaux de reprographie",
        "Gestion des stocks: Contrôler l'état des stocks",
        "Gestion des stocks: Définir des besoins en approvisionnement",
        "Logistique: Organiser le traitement des commandes",
        "Relation client: Accueillir, orienter, informer une personne",
        "Organisation: Contrôler la conformité des données ou des documents",
        "Organisation: Corriger et mettre en forme un document",
        "Organisation: Numériser des documents, médias ou supports techniques",
        "Organisation: Utiliser les outils bureautiques",
        "Organisation: Établir, mettre à jour un dossier, une base de données",
      ],
    },
    "workplace": {
      "brand": null,
      "description": null,
      "domain": {
        "idcc": 1000,
        "naf": null,
        "opco": "ATLAS",
      },
      "legal_name": null,
      "location": {
        "address": "3 rue du loup 75001 Paris",
        "geopoint": {
          "coordinates": [
            0,
            0,
          ],
          "type": "Point",
        },
      },
      "name": null,
      "siret": "11000001500013",
      "size": null,
      "website": null,
    },
  },
]
`;
